src/diem/testing/miniwallet/app/app.py

```diff
@@ -176,77 +196,48 @@ def _send_internal_payment_txn(self, txn: Transaction) -> None:
-                    sender_address=self.diem_account.account_identifier(),
-                    receiver=str(txn.payee),
-                    reference_id=reference_id,
-                    counterparty_account_identifier=identifier.encode_account(
-                        diem_id_address, None, self.diem_account.hrp
-                    ),
-                    sign=self.diem_account.sign_by_compliance_key,
-                )
-                receiver_address, _ = self.diem_account.decode_account_identifier(
-                    response.result.get("receiver_address")  # pyre-ignore
-                )
-                self.store.update(txn, payee_onchain_address=receiver_address.to_hex())
-            except offchain.Error as e:
-                if e.obj.code == ErrorCode.duplicate_reference_id:
-                    return
         if await self.offchain.is_under_dual_attestation_limit(txn.currency, txn.amount):
-            if not txn.signed_transaction:
-                signed_txn = await self.diem_account.submit_p2p(txn, await self.txn_metadata(txn))
-                self.store.update(txn, signed_transaction=signed_txn)
+            await self.send_diem_p2p_transaction(txn)
         else:
             await self.dual_attestation_txn_senders["v2"](txn)
 
-    async def _create_transaction(
-        self,
-        account_id: str,
-        status: str,
-        data: JsonInput,
-        type: Transaction.Type,
-        payee: Optional[str] = None,
-        subaddress_hex: Optional[str] = None,
-    ) -> Transaction:
-        return self.store.create(
-            Transaction,
-            account_id=account_id,
-            currency=await data.get("currency", str, self._validate_currency_code),
-            amount=await data.get("amount", int, self._validate_amount),
-            payee=payee,
-            status=status,
-            type=type,
-            subaddress_hex=subaddress_hex,
-            before_create=self._validate_account_balance,
+    async def send_diem_p2p_transaction(self, txn: Transaction, address: Optional[AccountAddress] = None) -> None:
+        if not txn.signed_transaction:
+            signed_txn = await self.diem_account.submit_p2p(txn, await self.txn_metadata(txn), by_address=address)
+            self.store.update(txn, signed_transaction=signed_txn.bcs_serialize().hex())
+        await self._wait_for_diem_txn(txn)
+
+    async def _wait_for_diem_txn(self, txn: Transaction) -> None:
+        try:
+            diem_txn = await self.diem_client.wait_for_transaction(str(txn.signed_transaction), timeout_secs=0.1)
+            self.store.update(txn, status=Transaction.Status.completed, diem_transaction_version=diem_txn.version)
+        except jsonrpc.WaitForTransactionTimeout as e:
+            self.logger.info("wait for txn(%s) timeout: %s, re-check later", txn, e)
+        except jsonrpc.TransactionHashMismatchError as e:
+            self.logger.warn("txn(%s) hash mismatched(%s), re-send diem p2p txn", txn, e)
```

Comment:
sorry for my not understanding, but could this happen during this flow?

Reply:
it happens when you submit 2 transactions for same account with same account sequence.
as async running in concurrent, it is possible `self.diem_account.submit_p2p` is called twice which will get latest account sequence number for submitting the txn, then going into `_wait_for_diem_txn`.

Reply:
I intent to not manage account sequence number in app, always just get latest account sequence number from account when submitting the transaction. 
Managing account sequence number is a harder problem (need make sure every sequence number is executed, so that no txn with bigger sequence number will be blocked) comparing with just get latest one and retry if we get hash mismatch error.

Reply:
awesome, thank you :-)